### PR TITLE
Separate CTE materialization session Properties from exchange Materialization

### PR DESCRIPTION
**Pull request by jaystarshot**
1. Enchances control of cte materialization hash partition buckets. Previously same property as exchanges were used

Current testing has shown that recommended value is 4x but further performance testing might be needed.

Meta recommends setting hash partition for materialized exchange as 5X - 10X the cluster size in exchange materialization and we do not want that as the hash partition count of non materialized exchanges

https://prestodb.io/docs/current/admin/exchange-materialization.html#using-exchange-materialization

2. Separates cte_partitioning_provider_catalog from exchange materialization 

This is needed since its good separation between features since these are cross used 

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

General Changes
* Added CTE hash partition count session property

### CODE REVIEW

**Review by feilong-liu — APPROVED** [review]

**Review by tdcmeehan — APPROVED** [review]

**Comment by steveburnett** [comment]
Could documentation for this new session property be added? Possibly in [Properties Reference](https://prestodb.io/docs/0.284/admin/properties.html).

File: presto-docs/src/main/sphinx/admin/properties.rst [hunk 1]
```diff
@@ -845,3 +845,26 @@ The following properties allow tuning the :doc:`/functions/regexp`.
...
+--------------------------------------
+
+``query.cte-hash-partition-count``
+^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Default value:** ``100``
+
+    The number of partitions to be used for materializing Common Table Expressions (CTEs) in queries.
+    This setting determines how many buckets/writers should be used when materializing the CTEs, potentially affecting the performance of queries involving CTE materialization.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties.rst:859]:
```suggestion
    This setting determines how many buckets or writers should be used when materializing the CTEs, potentially affecting the performance of queries involving CTE materialization.
```
Avoid slashes. See [GitLab doc recommended word list entry for slashes](https://docs.gitlab.com/ee/development/documentation/styleguide/word_list.html#slashes) for advice and suggestions.

**Review by steveburnett — CHANGES_REQUESTED** [review]
Thanks for the documentation! One nit only.

**Comment by jaystarshot** [comment]
Added the docs, but isn't this page redundant since we already have @ConfigDescription? Does it make more sense to add @ConfigDescription to all new session properties and pull them in the docs page during releases?

**Comment by steveburnett** [comment]
> Added the docs, but isn't this page redundant since we already have @ConfigDescription? Does it make more sense to add @ConfigDescription to all new session properties and pull them in the docs page during releases?

That's a great idea! I'll look into this possibility for generating the docs. Thanks very much for bringing it up.

**Comment by jaystarshot** [comment]
@steveburnett  Added documentation for cte-materialization-strategy property, please review

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

Pull branch, local build of docs, looks good. Thanks!

**Review by tdcmeehan — CHANGES_REQUESTED** [review]
LGTM % nit on the first commit, and please also add session property documentation.

Please make sure it follows our guidelines.  Suggested:

```
Add cte_hash_partition_count and config property
This commit adds the `cte_hash_partition_count`, and corresponding config property `query.cte-hash-partition-count`, which controls number of writers for CTE materialization effectively
```

File: presto-docs/src/main/sphinx/admin/properties.rst [hunk 2]
```diff
@@ -845,3 +845,37 @@ The following properties allow tuning the :doc:`/functions/regexp`.
...
+    A higher number of partitions might improve parallelism but also increases overhead in terms of memory and network communication.
+    Recommended value: 4 - 10x times the size of the cluster
+
+``query.cte-partitioning-provider-catalog``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``string``
+    * **Default value:** ``system``
+
+    The name of the catalog to be used for cte materialization and which provides custom partitioning for Common Table Expression (CTE) materialization.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties.rst:880]:
```suggestion
    The name of the catalog to be used for Common Table Expressions (CTE) materialization and which provides custom partitioning for CTE materialization.
```

File: presto-docs/src/main/sphinx/admin/properties.rst [hunk 3]
```diff
@@ -845,3 +845,37 @@ The following properties allow tuning the :doc:`/functions/regexp`.
...
+    This setting specifies which catalog should be used for cte materialization and for determining how to partition the materialization of CTEs in queries.
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties.rst:881]:
```suggestion
    This setting specifies which catalog should be used for CTE materialization and for determining how to partition the materialization of CTEs in queries.
```

**Review by steveburnett — CHANGES_REQUESTED** [review]
Took another look, two minor suggestions.

File: presto-docs/src/main/sphinx/admin/properties.rst [hunk 4]
```diff
@@ -845,3 +845,40 @@ The following properties allow tuning the :doc:`/functions/regexp`.
...
+``query.cte-hash-partition-count``
+^^^^^^^^^^^^^^^^^^^^
+
+    * **Type:** ``integer``
+    * **Default value:** ``100``
+
+    The number of partitions to be used for materializing Common Table Expressions (CTEs) in queries.
+    This setting determines how many buckets or writers should be used when materializing the CTEs, potentially affecting the performance of queries involving CTE materialization.
+    A higher number of partitions might improve parallelism but also increases overhead in terms of memory and network communication.
+    Recommended value: 4 - 10x times the size of the cluster
```
Comment (steveburnett) [inline presto-docs/src/main/sphinx/admin/properties.rst:873]:
```suggestion
    Recommended value: 4 - 10x times the size of the cluster.
```
Without the period the paragraph looks like this in a new build
Screenshot 2024-01-22 at 4 47 32 PM

**Review by steveburnett — CHANGES_REQUESTED** [review]
The tiniest of nits.

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)

Pull branch, local build of docs.

**Review by tdcmeehan — APPROVED** [review]

**Pull request merged**